File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/recovery/EntityContent.kt
package stasis.client_android.lib.ops.recovery

import okio.Source
import stasis.client_android.lib.api.clients.Clients
import stasis.client_android.lib.compression.Compression
import stasis.client_android.lib.encryption.Decoder
import stasis.client_android.lib.encryption.secrets.DeviceSecret
import stasis.client_android.lib.model.EntityMetadata
import stasis.client_android.lib.ops.recovery.stages.EntityProcessing
import stasis.client_android.lib.ops.recovery.stages.internal.DecompressedSource.decompress
import stasis.client_android.lib.ops.recovery.stages.internal.MergedCrates.merged
import java.nio.file.FileSystems

object EntityContent {
    suspend fun pull(
        metadata: EntityMetadata.WithContent,
        entityKey: String,
        deviceSecret: DeviceSecret,
        clients: Clients,
        decryptor: Decoder,
        onPartProcessed: () -> Unit
    ): Source {
        val crates = metadata.crates
        val filesystem = FileSystems.getDefault()

        val parts = crates.map { (partPath, crate) ->
            val decrypted = suspend {
                val encrypted = when (val source = clients.core.pull(crate)) {
                    null -> throw RuntimeException("Failed to pull crate [$crate] for entity [$entityKey]")
                    else -> source
                }

                decryptor.decrypt(encrypted, deviceSecret.toFileSecret(partPath, metadata.checksum))
            }

            Triple(EntityProcessing.partIdFromPath(partPath, filesystem), partPath, decrypted)
        }

        val lastPartId = parts.maxOfOrNull { it.first } ?: 0
        require(lastPartId + 1 == crates.size) {
            "Unexpected last part ID [$lastPartId] encountered for an entity with [${crates.size}] crate(s)"
        }

        return parts
            .merged(onPartProcessed = onPartProcessed)
            .decompress(decompressor = Compression.fromString(metadata.compression))
    }
}

File: client-ios/lib/Sources/StasisClientLib/Ops/Recovery/EntityContent.swift
import Foundation

public enum EntityContent {
    public static func pull(
        metadata: any EntityContentMetadata,
        entityKey: String,
        deviceSecret: DeviceSecret,
        clients: any Clients,
        decryptor: any Decrypting,
        onPartProcessed: @escaping @Sendable () async -> Void
    ) async throws -> AsyncThrowingStream<Data, Error> {
        let pulled = try await pullCrates(crates: metadata.crates, entityKey: entityKey, clients: clients)
        let decrypted = DecryptedCrates.decrypt(
            pulled,
            withPartSecret: { partPath in
                deviceSecret.toFileSecret(forFile: partPath, checksum: metadata.checksum)
            },
            decryptor: decryptor
        )
        let merged = try MergedCrates.merge(decrypted, onPartProcessed: onPartProcessed)
        let decompressor = try Compressions.fromString(metadata.compression)
        return DecompressedSource.decompress(merged, decompressor: decompressor)
    }

    public static func pullBytes(
        metadata: any EntityContentMetadata,
        entityKey: String,
        deviceSecret: DeviceSecret,
        clients: any Clients,
        decryptor: any Decrypting,
        onPartProcessed: @escaping @Sendable () async -> Void
    ) async throws -> Data {
        let stream = try await pull(
            metadata: metadata,
            entityKey: entityKey,
            deviceSecret: deviceSecret,
            clients: clients,
            decryptor: decryptor,
            onPartProcessed: onPartProcessed
        )
        var data = Data()
        for try await chunk in stream {
            data.append(chunk)
        }
        return data
    }

    private static func pullCrates(
        crates: [String: CrateId],
        entityKey: String,
        clients: any Clients
    ) async throws -> [RecoveryCrate] {
        let core = try await clients.core()
        let recovered: [RecoveryCrate] = crates.map { partPath, crate in
            let partId = Recovery.EntityProcessing.partIdFromPath(partPath)
            return RecoveryCrate(partId: partId, partPath: partPath) {
                guard let data = try await core.pull(crate: crate) else {
                    throw RecoveryPullError.crateMissing(crate: crate, entity: entityKey)
                }
                return AsyncThrowingStream { continuation in
                    continuation.yield(data)
                    continuation.finish()
                }
            }
        }
        let lastPartId = recovered.map(\.partId).max() ?? 0
        guard lastPartId + 1 == crates.count else {
            throw RecoveryPullError.unexpectedLastPartId(lastPartId: lastPartId, crateCount: crates.count)
        }
        return recovered
    }
}
